examples/schrodinger.dx

```diff
@@ -0,0 +1,143 @@
+'# 2D Time-Dependent Schrodinger Equation
+
+' The time-dependent Schrodinger equation in 2D is given by:\
+$-\frac{\hbar^2}{2m}(\frac{\partial^2}{\partial x^2} + \frac{\partial^2}{\partial y^2})\psi(x, y, t) + V(x, y)\psi(x, y, t) =
+i\hbar\frac{\partial\psi(x, y, t)}{\partial t}$
+
+' Setting $\hbar=m=1$ gives:\
+$-(\frac{\partial^2}{\partial x^2} + \frac{\partial^2}{\partial y^2})\psi(x, y, t) + V(x, y)\psi(x, y, t) =
+i\frac{\partial\psi(x, y, t)}{\partial t}$
+
+' Here we approximate this for small $\Delta x$ and $\Delta t$ by the finite difference method in two dimensions:\
+$$ \psi(x, y, t + \Delta t) = \\\\ \psi(x, y, t) \\\\ + \frac{i}{2}\frac{\Delta t}{\Delta x^2}(\psi(x-\Delta x, y, t) + \psi(x+\Delta x, y, t) + 4\psi(x, y, t) + \psi(x, y-\Delta y, t) + \psi(x, y+\Delta y, t)) \\\\ - i \Delta t V(x, y)\psi(x, y, t) $$
+
+'## Imports
+
+import plot
+import png
+
+'## Parameters
+
+-- The number of points on the grid in each dimension e.g. (41x41 for the 2d case)
+gridSize = 41
+D = Fin gridSize
+
+-- Space discretization
+dx = 1. / IToF gridSize
+
+-- Time discretization (solution unstable if much higher)
+dt = 0.000001
+
+-- The number of dt timesteps to simulate
+Steps = Fin 155000
+
+-- The number of frames to output in each GIF animation
+gifFrames = 100
+
+'## Helpers
+
+-- Define i_=sqrt(-1) in a way that won't clash with indices named 'i'
+i_ = MkComplex 0. 1.
+
+-- Shorthand conversion
+def fToC (x:Float) : Complex = MkComplex x 0.
+
+-- Translate from index to Float representation of real space
+-- Simulates the range [0.0, 1.0] in each dimension
+def ixToF (i:n) : Float = (IToF (ordinal i)) / (IToF gridSize - 1.)
+
+-- Helper to generate zero-delay GIF animations
+def animate (imgs:t=>n=>m=>(Fin 3)=>Float) : Html =
+  imgToHtml $ pngsToGif 0 $ map imgToPng imgs
+
+-- Cuts an array down to a smaller dimension.
+-- Useful for cutting out animation frames for a faster GIF.
+def cut (x:n=>a) : m=>a =
+  s = idiv (size n) (size m)
+  for i:m. x.(unsafeFromOrdinal _ $ s * ordinal i)
+
+-- Converts a given 2D matrix to a greyscale image.
+def toImg (xs:n=>m=>Float) : (n=>m=>Fin 3=>Float) =
+  scale = 1. / maximum (map maximum xs)
+  for h.
+    for w c. scale * xs.h.w
+
+'## Computing $\psi_{t+\Delta t}$
+
+-- Are the given indices at the bounds of the grid
+def atBounds (i:n) (j:m) : Bool =
+  (ordinal i == 0
+   || ordinal i == (gridSize - 1)
+   || ordinal j == 0
+   || ordinal j == (gridSize - 1))
+
+-- Run a single forward-step to compute psi_(t+1) from psi_t with the given potential.
+def step (psi:m=>n=>Complex) (v:m=>n=>Float) : m=>n=>Complex =
+  for i j.
+    -- Consider the manhattan nearest neighbours
+    -- Safe at the bounds due to the runtime bounds-check below
+    psi_u = psi.(unsafeFromOrdinal _ $ ordinal i - 1).j
+    psi_d = psi.(unsafeFromOrdinal _ $ ordinal i + 1).j
```

Comment:
Check out the `+|` operator from `benchmarks/parboil/stencil.dx`. I generally recommend packing unsafe casts into helpers. Since this one is actually more dangerous you could call it `+!` or something like that.

Reply:
That's a nice pattern, thanks. Changed to use of +! and -! operators.